src/MongoDB.Driver/Linq/Linq3Implementation/Serializers/KnownSerializers/KnownSerializerFinder.cs

```diff
@@ -22,33 +27,139 @@ internal class KnownSerializerFinder : ExpressionVisitor
     {
         #region static
         // public static methods
-        public static KnownSerializersRegistry FindKnownSerializers(Expression root)
+        public static KnownSerializersRegistry FindKnownSerializers(Expression root, IBsonDocumentSerializer rootSerializer)
         {
-            var visitor = new KnownSerializerFinder();
+            var visitor = new KnownSerializerFinder(root, rootSerializer);
             visitor.Visit(root);
             return visitor._registry;
         }
         #endregion
 
         // private fields
-        private KnownSerializersNode _expressionKnownSerializers = null;
-        private readonly KnownSerializersRegistry _registry = new KnownSerializersRegistry();
+        private KnownSerializersNode _currentKnownSerializersNode;
+        private IBsonDocumentSerializer _currentSerializer;
+        private readonly KnownSerializersRegistry _registry = new();
+        private readonly Expression _root;
+        private readonly IBsonDocumentSerializer _rootSerializer;
 
         // constructors
-        public KnownSerializerFinder()
+        private KnownSerializerFinder(Expression root, IBsonDocumentSerializer rootSerializer)
         {
+            _rootSerializer = rootSerializer;
+            _root = root;
         }
 
         // public methods
         public override Expression Visit(Expression node)
         {
-            _expressionKnownSerializers = new KnownSerializersNode(_expressionKnownSerializers);
-            _registry.Add(node, _expressionKnownSerializers);
+            if (node == null) return null;
+
+            _currentKnownSerializersNode = new KnownSerializersNode(_currentKnownSerializersNode);
+
+            if (node == _root)
+            {
+                _currentSerializer = _rootSerializer;
+            }
 
             var result = base.Visit(node);
+            _registry.Add(node, _currentKnownSerializersNode);
+            _currentKnownSerializersNode = _currentKnownSerializersNode.Parent;
+            return result;
+        }
+
+        protected override Expression VisitMember(MemberExpression node)
+        {
+            var result = base.VisitMember(node);
+            if (_currentSerializer.TryGetMemberSerializationInfo(node.Member.Name, out var memberSerializer))
+            {
+                PropagateToRoot(node, memberSerializer.Serializer);
+
+                if (memberSerializer.Serializer is IBsonDocumentSerializer bsonDocumentSerializer)
```

Comment:
Can we rename the out var to `documentSerializer`?

I realize `Bson` is in the interface name, but it seems redundant in the variable name.

Reply:
What if it's not a IBsonDocumentSerializer? Should we set the _currentSerializer to null then?